### ISSUE TITLE
Consider semver v2 for the master branch development version

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#14157]
We had a discussion on Slack regarding shifting the `VERSION` file for the master branch from the `1.X-1.90` style for development versions towards `1.X.0-dev`. For example, rather than using `1.9.90` version to denote a pre-v1.10 development version, instead use `1.10.0-dev`. This would be the version number tagged on "nightly" type images, such as `latest` docker images and pre-release helm charts.

## Positives

* Mitigate confusion for projects that vendor cilium (eg hubble-cli)
  * Eg. right now hubble-cli has this in the go.mod:
    `github.com/cilium/cilium v1.9.0-rc1.0.20201120102040-c0b68419e6d8`
    This looks like a pre-1.9 version, but it's actually just the latest tag in the master branch and the actual code is a 1.10 development version.
* Reduce likelihood that someone could deploy a latest version and confuse "1.9.9" vs. 1.9.90"
  * Not so likely since the user would have to install `cilium/cilium:latest` which users are usually steered away from in the docs.
* Easier to generate helm charts for the master branch
  * Rather than modifying the `VERSION` file in the repo, it would already have the correct version.
  * Note, if we want to update older branch pre-release helm charts, we still need to manually modify the version in the tree.

## Concerns

* Possibility to break something unintended in the tree
  * Related:
  ```
  $ git grep "\.90"
  Documentation/concepts/kubernetes/troubleshooting.rst:        INFO Cilium 0.8.90 f022e2f Thu, 27 Apr 2017 23:17:56 -0700 go version go1.7.5 linux/amd64
  Documentation/contributing/release/feature.rst:#. Update the ``VERSION`` file to contain ``v1.2.90``
  Documentation/gettingstarted/aws.rst:          lastUpdated: 2018-11-30T11:28:03.907678888Z
  Binary file Documentation/gettingstarted/images/cilium_k8s_demo_l7-policy-230817.png matches
  Binary file Documentation/images/identity_store.png matches
  VERSION:1.9.90
  clustermesh-apiserver/test/mock.json:{"kind": "Endpoints", "metadata": {"name": "foo", "namespace": "default"}, "subsets": [{"addresses": [{"ip": "34.90.138.138"}], "ports": [{"name": "https", "port": 443, "protocol": "TCP"}]}]}
  contrib/packaging/deb/debian/changelog:cilium (0.9.90) UNRELEASED; urgency=medium
  contrib/packaging/deb/debian/changelog:  * Version 0.9.90
  contrib/packaging/rpm/cilium.spec.envsubst:* Wed Dec 20 2017 Tony Lambiris <tony@criticalstack.com> - 0.13.90-0.git${SHORTCOMMIT}
  install/kubernetes/Makefile:LATEST_VERSION_REGEX := '[0-9]\+\.[0-9]\+\.90'
  install/kubernetes/cilium/Chart.yaml:version: 1.9.90
  install/kubernetes/cilium/Chart.yaml:appVersion: 1.9.90
  install/kubernetes/cilium/README.md:![Version: 1.9.90](https://img.shields.io/badge/Version-1.9.90-informational?style=flat-square) ![AppVersion: 1.9.90](https://img.shields.io/badge/AppVersion-1.9.90-informational?style=flat-square)
  pkg/maps/ctmap/ctmap.go:                // 25%..90% => 1.3x..10x shorter
  pkg/version/version.go: // string to parse: "0.13.90 a722bdb 2018-01-09T22:32:37+01:00 go version go1.9 linux/amd64"
  pkg/version/version_test.go:                    "0.11.90 774ecd3 2018-01-09T22:32:37+01:00 go version go1.8.3 linux/amd64",
  pkg/version/version_test.go:                            Version:          "0.11.90",
  pkg/version/version_test.go:                    "0.11.90 774ecd3 2018-01-09T22:32:37+01:00 go version go1.9 someArch/i8726",
  pkg/version/version_test.go:                            Version:          "0.11.90",
  pkg/version/version_test.go:                    "0.13.90 7330b8d 2018-01-09T22:32:37+01:00 go version go1.8.3 linux/arm",
  pkg/version/version_test.go:                            Version:          "0.13.90",
  pkg/version/version_test.go:                    "0.13.90 7330b8d linux/arm",
  test/helpers/cons.go:   CiliumLatestHelmChartVersion = "1.8.90"
  test/helpers/cons.go:   IsCiliumV1_5 = versioncheck.MustCompile(">=1.4.90 <1.6.0")
  test/helpers/cons.go:   IsCiliumV1_6 = versioncheck.MustCompile(">=1.5.90 <1.7.0")
  test/helpers/cons.go:   IsCiliumV1_7 = versioncheck.MustCompile(">=1.6.90 <1.8.0")
  test/helpers/cons.go:   IsCiliumV1_8 = versioncheck.MustCompile(">=1.7.90 <1.9.0")
  test/helpers/cons.go:   IsCiliumV1_9 = versioncheck.MustCompile(">=1.8.90 <1.10.0")
  ```

**Comment by ti-mo** [comment]
This is still the case today, are we still planning on changing it? If so, what's the estimated effort and who should pick it up?

**Comment by joestringer** [comment]
We already did this. During v1.14 we also toyed with `X.Y.Z-snapshot.N` formatting for intermediate releases (mostly monthly).

I think the only next thing to evaluate for v1.15 is that during v1.15 we identified that the `-snapshot.N` suffix would cause complications for semantic version evaluation when we start creating RCs due to `X.Y.Z-rc.N` being semantically earlier than `X.Y.Z-snapshot.N`. I plan to propose to use something like `X.Y.Z-pre.N` for prerelease snapshots in the next cycle to mitigate this issue.
